STYLEGUIDE.md

```diff
@@ -0,0 +1,172 @@
+too large. In these cases, we add a line break right after the `->`-symbol
+following the lambda-abstraction, which is the earliest branching point in this
+case. The next node is again `Σ`, with two arguments. The first one fits on a
+line, but the second does not, so we add a line break between them. This process
+is continued until the definition is complete.
+
+Note also that we use parentheses to mark the branches. The extra space after
+the opening parentheses marking a branch is there to visually emphasize the tree
+structure of the definition, and synergizes with our convention to have
+two-space indentation level increases.
+
+## Naming conventions
+
+- As a main strategy, we strive to keep a tight connection between names of
+  constructions and their types. Take for instance [...].
+  > - Add example
+  > - prepending assumptions and then conclusion
+  > - See agda-unimath's description
+  > - > We start with the initial assumption, then, working our way to the
+  >   > conclusion, prepending every central assumption to the name, and finally
+  >   > the conclusion. So for instance the name `iso-is-initial-is-segal`
+  >   > should read like we get an iso of something which is initial given that
+  >   > something is Segal. The true reading should then be obvious.
+  >
+  > > The naming conventions are aimed at improving the readability of the code,
+  > > not to ensure the shortest possible names, nor to minimize the amount of
+  > > typing by the implementers of the library.
+- We mainly use lower case names with words separated by hyphens.
+- Capitalized names are reserved for subuniverses and similar collections. When
+  a construction is made internally to such a collection, we _append_ its name.
+  For instance, the subuniverse of Segal types is called `Segal`, and its
+  internal hom, called `function-type-Segal,` has the following signature:
+
+  ```rzk
+  #def function-type-Segal
+    ( A B : Segal)
+    : Segal
+  ```
+
+- Use meaningful names that accurately represent the concepts applied. For
+  example, if a concept is known best by a special name, that name should
+  probably be used.
+
+- For technical lemmas or definitions, where the chance they will be reused is
+  very low, the specific names do not matter as much. In these cases, one may
+  resort to a simplified naming scheme, like enumeration. Please note, however,
+  that if you find yourself appealing to this convention frequently, that is a
+  sign that your code should be refactored.
+
+## Use of Comments
+
+> We do not explicitly ban code comments, but our other conventions should
+> heavily limit their need.
+>
+> - [ ] Literate file format for prose
+> - [ ] Descriptive definition names shouldn't need additional explanation
+> - [ ] Strictly organized code to ease reading and understanding
+> - [ ] Essential information should be carried by code, not only comments
+>
+> Still, code annotations may find their uses.
+>
+> Where to place literature references?
```

Comment:
Perhaps, we could use code block title? Example:

    The type of commutative triangles in `#!rzk A`:
    
    ```rzk title="RS17, Definition 5.2"
    #def hom2
      ( A : U)           -- A type.
      ( x y z : A)       -- Three points in A.
      ( f : hom A x y)   -- An arrow in A from x to y.
      ( g : hom A y z)   -- An arrow in A from y to z.
      ( h : hom A x z)   -- An arrow in A from x to z.
      : U                -- (hom2 A x y z f g h) is a 2-simplex (triangle)
      := { (t1, t2) : Δ² } -> A [   -- in A where
        t2 === 0_2 |-> f t1,        -- * the top edge is exactly f,
        t1 === 1_2 |-> g t2,        -- * the right edge is exactly g, and
        t2 === t1  |-> h t2         -- * the diagonal is exactly h
      ]
    ```

Rendered as the following (with MkDocs Material theme):
<img width="812" alt="Screenshot 2023-06-30 at 19 24 05" src="https://github.com/emilyriehl/yoneda/assets/686582/7e1e876c-ba74-41a3-af22-3d6b15f00e63">

Unfortunately, titles do not support rich text, so no links/footnotes available.

Reply:
By the way, what is the agda-unimath way to format this definition after `:=`? :)

Reply:
I like this use of titles for now. @fizruk can you add it to the style guide?

Reply:
Perhaps something like this, although I'm not entirely content with this suggestion.

```rzk title="RS17, Definition 5.2"
#def is-composition
  ( A : U)
  ( x y z : A)
  ( f : hom A x y)
  ( g : hom A y z)
  ( h : hom A x z)
  : U
  :=
    Δ² ->
    A [ t2 === 0_2 |-> f t1 ,
        t1 === 1_2 |-> g t2 ,
        t2 === t1  |-> h t2]
```

Reply:
I would also prefer using unicode alternatives for `->`, `===` and `|->`.

Reply:
Yes, I will get Unicode versions of these back in one of the next releases, then we can fully switch to that.

Reply:
#13 switches to Material theme which allows titles for code blocks among other visual things. If Material does not look good, I can try bringing titles to the current theme (apparently there is no out-of-the-box solution for the current theme).

Reply:
> I would also prefer using unicode alternatives for `->`, `===` and `|->`.

Unicode syntax support is in `develop` (see https://github.com/fizruk/rzk/pull/61) and will be out with the next release of `rzk`.

Reply:
@fizruk  On second thought, your example should be formatted more like this:

```rzk
#def is-composition
  ( A : U)
  ( x y z : A)
  ( f : hom A x y)
  ( g : hom A y z)
  ( h : hom A x z)
  : U
  :=
    Δ² -> A
    [ t2 === 0_2 |-> f t1 ,
      t1 === 1_2 |-> g t2 ,
      t2 === t1  |-> h t2]
 ```

Reply:
I'm still not entirely decided on this though, and I'm still very much in the process of learning the language. Every input is much appreciated!

Reply:
What do you think of the suggestion at https://github.com/emilyriehl/yoneda/issues/14#issuecomment-1618104985 ? Note that the language is still a prototype and can be adjusted.

Reply:
Oh, I'm not sure yet. I need some more time to get used to the syntax of `rzk`. It looks promising, though I think it is important that there aren't too many different notations for the same construction. I think two notations should be fine, though.